### PR TITLE
Report diff coverage for the lines a branch changed

### PR DESCRIPTION
**Pull request by vlsi**
The build collects JaCoCo data only under -PenableJacoco, generates no per-module report, and aggregates through :code-coverage-report, whose coverageDataElementsForTest wiring pulls in testJdk17, testJdk21, testJdk26 and testJdk28. Coverage was therefore reachable only by re-running the suite on five JDKs, which is too expensive to consult while writing a test, so nobody consulted it and a test that never reached the code it was written for looked the same as one that did.

Instrument the `test` task by default and write its XML report as a finalizer of that task. Measured on :nullaway:test, 1098 tests: instrumentation takes the run from 16.4 s to 20.0 s and the report adds about 1 s. -PdiffCoverageEnabled=false takes back everything the default turns on, the report, the XML it reads and the instrumentation that exists only to produce them, so that one flag buys the 20% back for a run that is not about coverage. -PenableJacoco still decides instrumentation wherever it is set, off or extended to the testJdk* tasks for the aggregate report; CI already sets the property explicitly on every platform, so the new default does not change what CI runs.

The diffCoverage task intersects that report with the lines git diff marks as changed under the module's main source roots, and prints the ones the run did not execute and the branches it took one way. It compares no stored coverage data: the base is a git ref whose merge base with HEAD decides which lines count as changed, so nothing has to be recorded on the base commit or committed to the repository. A ratchet against a stored baseline was rejected because it needs a second instrumented run of the base tree and answers a worse question, reporting that coverage fell by 0.1% rather than naming a line. The task prints at the QUIET log level, so --quiet still shows it, and it names the ref that resolved and the test filter the counts came from, since a filtered run leaves the rest of the change looking unexecuted.

The report is advisory and fails no build. Failing to resolve the base ref or to find a report prints a message saying so, rather than the "no changed executable lines" that would read as "everything is covered". A changed file no report covers is named in a block of its own for the same reason: it counts towards neither side of the total, so a change made entirely in such files would otherwise print as fully covered without having been measured. An untracked file this JVM cannot decode is named and left out instead, since failing a build whose tests passed over an advisory report is the trade the other way round.

Every git command the task runs asks about the repository rather than about the module: the base ref, the diff of every Java file, the untracked files. A shared build service runs them, keyed by the whole argument list, so twelve modules produce one run of each command rather than twelve, and a working tree edited while the build runs cannot make two modules print different changed lines for one change. A command that git rejected is cached with its standard error, since a failure and a command that succeeded and found nothing must not read alike.

The finalizer is wired only where nobody set -PenableJacoco and nobody turned the report off: every CI job that reaches jacocoTestReport sets the property, and a CI checkout fetches the pull request ref alone, so there the task would find no base ref and say so once per module on every build. The conventions plugin gives all twelve modules the task, so a whole-build run prints a block for each module the change touched and nothing for the others. Whether the change reached the module is settled before the report is looked for, because a module with no test of its own produces no report, and :annotations would otherwise report a missing one on every whole-build run.

Measured on a full `./gradlew build --rerun-tasks --no-build-cache`, twelve modules with the JDK matrix: 167 s with -PenableJacoco=false, about 180 s on the new default, and 216 s with -PenableJacoco=true, which instruments the testJdk17, testJdk21, testJdk26 and testJdk28 runs as well. The default therefore costs around 8% of a full build and the aggregate configuration around 30%.

The base is the first ref that resolves of `-PdiffCoverageBase`, the upstream branch, and `origin/master`, and the diff is taken from its merge base with HEAD. A ref that resolves and shares no commit with HEAD, as in a shallow clone or after an orphan branch, gives no merge base and counts as no base at all, rather than failing a build whose tests passed. An upstream that is this branch's own pushed copy is left out: in a fork workflow the branch tracks the fork, so after the first push its upstream holds every line the branch changed, and the report would say nothing was changed rather than that it could not find a base. The remote name is dropped before that comparison, so a branch named after the tail of another one keeps its own upstream.

A diff line counts as a file header only inside the header section git opens with `diff --git`, and only as the `+++` directly under its `---`. Content carries both shapes: an added line reading `++ x` arrives as `+++ x`, and replacing a line `-- a/x` with a line `++ b/y` arrives as a whole well-formed header pair. Reading either as a header keys the rest of that file's hunks under a path no source root claims, which drops those lines from the report with nothing to show that it happened. Where two source roots are nested, the longer one keys the file, since the shorter one would leave its extra directories in the package path and match no JaCoCo entry. Both git commands run with core.quotePath off, since a path holding a non-ASCII byte otherwise arrives quoted and escaped, as `"b/src/café.java"`: such a `+++` line starts with a quotation mark rather than with the prefix, so it is no file header and every hunk of that file leaves the report silently, and the same path from ls-files names no file this JVM can open.

Only the files that still owe a test are described, so that the block reads as the work left rather than as an inventory of the change; the fully covered ones are counted in one line, which is what keeps "measured and fine" apart from "never measured". They are ordered by uncovered changed lines, then by uncovered branch outcomes, then by how much the file changed, then by path. The keys stay separate rather than summing into a score, because a line and a branch outcome are different things and any coefficient joining them would be invented, and the counts are absolute rather than shares, because the question is where the most work is left: 90 of 100 lines executed leaves ten lines to test and 1 of 2 leaves one, however much worse the second reads as a percentage. A filtered run says that the rest of the change looks unexecuted, because the scope alone has not stopped anyone quoting its total as the branch's.

Each file gets one row per uncovered line, carrying the line number, why the line counts as uncovered, and the source stripped and cut to 100 characters. The quote is what makes a row actionable with no editor open beside it. A line the run neither reached nor took a branch of is one line of work, so `194: not executed; 0 of 2 branches covered` is one row rather than an entry in each of two lists: reported twice, the line inflates every count that adds the rows up, and the branch list was a subset of the pair above it that a reader who added them up found inconsistent.

The block describes at most 10 files, at most 10 rows per file and at most 50 rows over its lists together, and -PdiffCoverageMaxFiles, -PdiffCoverageMaxLinesPerFile and -PdiffCoverageMaxLines move those limits. One budget covers the whole block rather than one per list, so that a block capped at 50 rows is 50 rows; the work is printed first, and the acknowledged lines below get what it leaves. A truncated list states its own total, and -PdiffCoverageShowAll prints the report whole.

What the limits leave out is in <module>/build/reports/diff-coverage/test.txt, whose path the block prints beside the JaCoCo XML's, so that reading the rest costs a file read rather than a parse of the XML. A run that measures nothing deletes that file, since a reader holding the path from an earlier run would otherwise read that run's report as this one's.

A comment acknowledges a line nothing covers: `throw new AssertionError(); // diff-coverage: ignore -- defensive safeguard` moves the line into a list of its own, so that the next reader does not weigh a line somebody has already decided about. It changes no count: the line still counts as unexecuted in the total, and `Uncovered changed code: 27 lines (1 acknowledged)` names how many of them carry a marker. The marker has to follow a comment opener on the line it acknowledges; nothing here parses Java, so a marker whose opener sits inside a string literal counts as well.

Each file's counts, and the total, carry branches beside lines. A line counts as executed once any path through it ran, so a changed condition tested one way reads as fully covered by line counts alone; the branch pair is what reports it. A change whose lines branch nowhere gets no branch pair rather than a 0/0 one.

Running the module's suite with -PdiffCoverageBase=ebd9b822~1, which makes the lines of the four commits since then count as changed, prints:

    Diff coverage against ebd9b822~1 (e97f8ff91), from :nullaway:test
      ordered by uncovered changed lines, then uncovered branches

      nullaway/src/main/java/com/uber/nullaway/librarymodel/AddAnnotationToNestedTypeVisitor.java
        6/9 changed lines executed, 4/6 branches covered
        uncovered changed code:
          140: 3 of 4 branches covered: if (wildcard.kind == BoundKind.UNBOUND && wildcard.bound == null) {
          149: not executed: updatedWildcard = (Type.WildcardType) wildcard.accept(this, pathIndex);
          152: 1 of 2 branches covered: if (wildcard.kind == BoundKind.UNBOUND) {
          161: not executed: Type updatedBound = TypeSubstitutionUtils.typeWithAnnot(wildcard.type, annotationType);
          162: not executed: updatedWildcard = TYPE_METADATA_BUILDER.createWildcardType(wildcard, updatedBound);

      nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
        39/40 changed lines executed, 22/28 branches covered
        uncovered changed code:
          888: 5 of 6 branches covered: if (type != null && type.isRaw() && !(type instanceof Type.ArrayType)) {
          1087: not executed: return typeOrNullIfRawNonArray(symbol.type);
          1705: 5 of 6 branches covered: if (!config.isJSpecifyMode() || targetType.isRaw() || referencedMethod.isConstructor()) {
          1720: 3 of 4 branches covered: if (qualifierType == null || qualifierType.isRaw()) {
          1725: 1 of 2 branches covered: if (qualifierType instanceof Type.ClassType qualifierClassType) {
          1734: 1 of 2 branches covered: !fiParamTypes.isEmpty(),
          1745: 1 of 2 branches covered: return methodType == null ? null : new ResolvedMethodReference(methodType, qualifierType);

      [GenericsUtils.java and NullAway.java, abridged here]

      nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
        1/1 changed lines executed, 1/2 branches covered
        uncovered changed code:
          153: 1 of 2 branches covered: Verify.verify(wildcard.kind == BoundKind.UNBOUND, "wildcard must be unbounded");

    Total: 78/82 changed lines executed, 47/60 branches covered.
    Uncovered changed code: 17 lines

    Full diff coverage report:
      nullaway/build/reports/diff-coverage/test.txt
    Full JaCoCo report:
      nullaway/build/reports/jacoco/test/jacocoTestReport.xml

The changed line of TypeSubstitutionUtils.java runs, so line counts alone report that file as covered; its branch row is what says the condition is exercised one way by the entire module suite. Running only GenericsTests against the same base reports 19/82 lines, and says under the header that a filtered run leaves the rest of the change looking unexecuted, which is the reading the scope line exists to prevent.

The buildSrc suite covers the format: the three row shapes, the quote at and past its width, each limit at and past its boundary, the one budget the two lists share, the marker and the four ways it does not fire, and the report file holding the rows the console left out. It takes a real diff of a path holding a non-ASCII byte, which reaches the parser as no file at all with core.quotePath left on. Removing the acknowledgement fails eight of those tests, giving the acknowledged list a budget of its own fails one, and cutting the quote a character early fails the boundary case. GitServiceTest covers the cache from both sides: a repeated command keeps the answer of its first run across a commit made between the two, and a command with different arguments does not.

Assisted-by: Claude Code (claude-opus-5)

### CODE REVIEW

File: buildSrc/src/main/groovy/nullaway.java-test-conventions.gradle [hunk 1]
```diff
@@ -39,21 +39,77 @@ jacoco {
...
+
+def diffCoverage = tasks.register("diffCoverage", com.uber.nullaway.gradle.DiffCoverageTask) {
+    description = "Reports which of the lines this branch changed the last test run executed"
+    group = LifecycleBasePlugin.VERIFICATION_GROUP
+    reportFiles.from(tasks.named("jacocoTestReport").map { it.reports.xml.outputLocation })
+    repositoryRoot = rootProject.layout.projectDirectory
+    sourceRoots.set(providers.provider {
+        sourceSets.main.java.srcDirs.collect { rootDir.toPath().relativize(it.toPath()).toString() }
+    })
+    // A filtered run leaves the rest of the change looking unexecuted, so the scope is printed
```
[the remark this answers is not in the document]
Comment (vlsi) [reply buildSrc/src/main/groovy/nullaway.java-test-conventions.gradle]: The roots do arrive with backslashes on Windows, and they are normalized where they are consumed: `DiffCoverageReport.sourceFileKey` runs `sourceRoot.replace('\\', '/')` before it matches a git path, and `DiffCoverageReportTest.aRootSpelledWithBackslashesStillMatchesAGitPath` passes `nullaway\src\main\java` and asserts the file is still measured.

That is the change that closed the same finding on line 89 of this file, so normalizing in the plugin as well would state the rule in two places, and the one that matters is the one next to the matching.

File: buildSrc/src/main/groovy/nullaway.java-test-conventions.gradle [hunk 2]
```diff
@@ -39,21 +39,77 @@ jacoco {
...
+    // beside the counts rather than left for the reader to remember.
+    def testFilterPatterns = providers.provider {
+        def testTask = tasks.getByName("test")
+        def patterns = testTask.filter.includePatterns as List
+        if (testTask.filter.metaClass.respondsTo(testTask.filter, "getCommandLineIncludePatterns")) {
+            patterns += testTask.filter.commandLineIncludePatterns as List
+        }
+        patterns
+    }
```
[the remark this answers is not in the document]
Comment (vlsi) [reply buildSrc/src/main/groovy/nullaway.java-test-conventions.gradle:105]: The providers do capture `tasks`, `project`, `sourceSets` and `rootDir`. This build does not enable the configuration cache, though: nothing in `gradle.properties`, in the workflows under `.github/`, or in the root scripts turns it on, and the other conventions plugins hold `project` the same way, so no run fails on this today.

Resolving the values at configuration time is also not a local change. Gradle sets `test.filter.commandLineIncludePatterns` when it selects the task from the command line, after this plugin has configured it, so reading the filter eagerly would drop `--tests` from the scope line — and that line is what keeps a filtered run from being read as missing coverage. Making the build configuration-cache compatible is worth a change of its own, across the conventions plugins rather than in this one.

File: buildSrc/src/test/java/com/uber/nullaway/gradle/DiffCoverageTaskTest.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+     * A run that measured nothing has to say so. Reusing the total a measured run prints for an
+     * empty change would tell the reader that everything they changed was covered.
+     */
+    @ParameterizedTest
+    @ValueSource(strings = {DiffCoverageTask.NO_BASE, DiffCoverageTask.NO_REPORT})
+    void aMessageSayingNothingWasMeasuredIsNotTheTotalOfAMeasuredRun(String message) {
+      assertFalse(
+          message.contains(DiffCoverageReport.NO_EXECUTABLE_LINES),
+          () -> "message reads as a measured run: " + message);
+    }
```
[the remark this answers is not in the document]
Comment (vlsi) [reply buildSrc/src/test/java/com/uber/nullaway/gradle/DiffCoverageTaskTest.java]: Each cause is already pinned where it is produced: `DiffCoverageTaskRunTest.WithNoBaseRef.aRunWithAThresholdFailsRatherThanPassingOnAnUnmeasuredChange` asserts `NO_BASE` for a repository with no commit, `WithNoReport.aRunWithAThresholdFailsRatherThanPassingOnAnUnmeasuredChange` asserts `NO_REPORT` for a run whose report is absent, and `WithNoCommonHistory.aBaseWithNoMergeBaseIsTreatedAsNoBaseAtAll` asserts `NO_BASE` for a ref that shares no commit with HEAD. Swapping the two constants at the call sites fails those three.

Asserting the wording here would pin the prose instead: the test would fail on a reworded message that reports the same cause, and it would still pass with the constants swapped, since this class runs no task. What it can establish without the wording is what it does establish — neither message can be read as the total of a measured run, and the two name different causes.

File: buildSrc/src/main/java/com/uber/nullaway/gradle/DiffCoverageTask.java [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+            + base.ref
+            + " ("
+            + base.commit.substring(0, Math.min(9, base.commit.length()))
+            + "), from "
+            + getScopeDescription().get();
+    List<String> stale = staleSources(repository, report.measuredPaths(), reports);
+    String text =
+        report.render(
+            scope, getFiltered().get(), stale, getMaxLines().get(), getMaxFiles().get());
+    failIfBelowThreshold(report);
```
[the remark this answers is not in the document]
Comment (vlsi) [reply buildSrc/src/main/java/com/uber/nullaway/gradle/DiffCoverageTask.java]: The threshold is gone as of 1ef3732a, so nothing thrown from `describeRun` can now swallow the block: `-PdiffCoverageFailUnder` had no user (no CI job set it, AGENTS.md never named it), and the report is advisory. `describeRun` returns the text and `report()` logs it; the only exception left is for an explicit `-PdiffCoverageBase` that does not resolve, which is a misconfiguration rather than a coverage result.

File: buildSrc/src/main/java/com/uber/nullaway/gradle/DiffCoverageReport.java [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+      if (!file.partlyTakenBranches.isEmpty()) {
+        // Without this the reader adds the rows up, finds fewer branches than the pair above
+        // reports, and concludes the block is broken. The lines whose branches were all covered
+        // are the difference, and they are not worth a row each.
+        out.append("\n    lines with an uncovered branch:");
+      }
+      int shownBranches =
+          file.partlyTakenBranches.size() > maxLines
+              ? maxLines - 1
+              : file.partlyTakenBranches.size();
```
[the remark this answers is not in the document]
Comment (vlsi) [reply buildSrc/src/main/java/com/uber/nullaway/gradle/DiffCoverageReport.java]: This describes the two-list rendering, where `shownBranches` was `maxLines - 1`. That split is gone: each uncovered line is one row, so `194: not executed; 0 of 2 branches covered` is a single entry rather than one in each list, and no subtraction feeds a `subList` endpoint. A limit of zero prints the file and none of its rows (`DiffCoverageReportTest.aLimitOfZeroPrintsTheFileAndNoneOfItsRows`); a negative one is refused by the `Limits` constructor with `a diff coverage limit cannot be negative: files=4, linesPerFile=-1, lines=50` (`aNegativeLimitIsRefusedWithAMessageNamingIt`).
